File: Anzac3/E/E.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

#define pb(a) push_back(a)
#define mp(a,b) make_pair(a,b)

using namespace std;

typedef long long ll;
typedef pair<ll, int> pli;

struct Chunk {
    ll a, b;
    bool is_single_char;
    char c;
    bool deleted;

    Chunk (ll x, ll y) : a(x), b(y), is_single_char(false), deleted(false) {};
    Chunk (char character) : c(character), is_single_char(true), deleted(false) {};

    ll size()
    {
        if (is_single_char) return 1;
        return b - a  + 1;
    }

    bool merge(Chunk other)
    {
        if (is_single_char || other.is_single_char) return false;
        if (b + 1 == other.a)
        {
            b = other.b;
            return true;
        }
        return false;
    }

    bool equals(Chunk other)
    {
        if (is_single_char != other.is_single_char) return 0;
        if (is_single_char) return c == other.c;
        return a == other.a && b == other.b;
    }
};

vector<Chunk> text[2];
const ll MAXN = 1e11;

pli find_pos_in_chunk(int index, ll pos)
{
    ll skipped = 0;
    for (int i = 0; i < text[index].size(); i++)
    {
        if (text[index][i].deleted) continue;
        if (skipped + text[index][i].size() >= pos)
            return mp(skipped, i);
        skipped += text[index][i].size();
    }
    // Shouldn't reach here
    int FAIL1 = 1;
    int FAIL2 = 2;
    return mp(0, FAIL1 / ((FAIL2 / 2) - 1));
}

void perform_delete(int index, ll pos)
{
    pli selected_chunk = find_pos_in_chunk(index, pos);
    ll skipped = selected_chunk.first;
    int chunk_id = selected_chunk.second;

    if (text[index][chunk_id].size() == 1)
        text[index][chunk_id].deleted = 1;
    else 
    {
        ll remaining = pos - skipped;
        ll a = text[index][chunk_id].a, b = text[index][chunk_id].b;
        
        Chunk left = Chunk(a, a + remaining - 2);
        Chunk right = Chunk(a + remaining, b);

        if (right.size() <= 0)
            text[index][chunk_id] = left;
        else if (left.size() <= 0)
            text[index][chunk_id] = right;
        else {
            text[index][chunk_id] = right;
            text[index].insert(text[index].begin() + chunk_id, left);
        }
    }
}

void perform_insert(int index, ll pos, char c)
{
    pli selected_chunk = find_pos_in_chunk(index, pos);
    ll remaining = pos - selected_chunk.first;
    int chunk_id = selected_chunk.second;

    Chunk new_chunk = Chunk(c);

    if (remaining == 1)
    {
        text[index].insert(text[index].begin() + chunk_id, new_chunk);
        return;
    }

    ll a = text[index][chunk_id].a, b = text[index][chunk_id].b;

    Chunk left = Chunk(a, a + remaining - 2);
    Chunk right = Chunk(a + remaining - 1, b);

    text[index][chunk_id].deleted = 1;

    if (right.size() >= 0)
        text[index].insert(text[index].begin() + chunk_id, right);
    text[index].insert(text[index].begin() + chunk_id, new_chunk);
    if (left.size() >= 0)
        text[index].insert(text[index].begin() + chunk_id, left);
}

void print(vector<Chunk> v)
{
    for (int i = 0; i < v.size(); i++)
    {
        Chunk chunk = v[i];
        if (chunk.deleted) continue;
        if (chunk.is_single_char) cout << chunk.c << ", ";
        else cout << "[" << chunk.a << ", " << chunk.b << "], ";
    }
    cout << endl;
}

void read_input(int index)
{
    char c;
    cin >> c;
    while (c != 'E')
    {
        if (c == 'D')
        {
            ll pos; cin >> pos;
            perform_delete(index, pos);
        }
        if (c == 'I')
        {
            ll pos; cin >> pos;
            char new_char; cin >> new_char;
            perform_insert(index, pos, new_char);
        }
        cin >> c;
    }
}

int main()
{
    text[0].pb(Chunk(1, MAXN));
    text[1].pb(Chunk(1, MAXN));

    read_input(0);
    read_input(1);

    vector<Chunk> merged[2];

    for (int text_id = 0; text_id < 2; text_id++)
    {
        for (int merging = 0; merging < text[text_id].size(); merging++)
        {
            if (text[text_id][merging].deleted) continue;
            if (merged[text_id].size() == 0 || !merged[text_id][merged[text_id].size()-1].merge(text[text_id][merging]))
                merged[text_id].pb(text[text_id][merging]);
        }
    }

    bool failed = merged[0].size() != merged[1].size();
    for (int i = 0; i < merged[0].size() && !failed; i++)
        failed |= !merged[0][i].equals(merged[1][i]);

    //print(merged[0]);
    //print(merged[1]);

    cout << (int) failed << endl;

}

File: Anzac3/E/E.py
class Chunk:
    def __init__(self, a, b=1e10 + 20000, single_char=False):
        if not single_char:
            self.a = int(a) # Inclusive
            self.b = int(b) # Inclusive
        if single_char:
            self.char = a
        self.single_char = single_char

    def __len__(self):
        return self.b - self.a + 1 if not self.single_char else 1

    def offset(self, val):
        return Chunk(self.a + val, self.b)

    def covering(self, other):
        if self.a != other.a:
            return 0
        return min(len(self), len(other))

    def __str__(self):
        if self.single_char:
            return self.char
        return "[{}:{}]".format(self.a, self.b)

    def merge(self, other):
        if self.single_char or other.single_char:
            return False
        if self.b in [other.a, other.a - 1]:
            self.b = other.b
            return True
        return False

    def equals(self, other):
        if self.single_char:
            return other.single_char and self.char == other.char
        if other.single_char:
            return False
        return self.a == other.a and self.b == other.b


def perform_delete(chunks, pos):
    skipped = 0
    selected_chunk = 0
    while skipped + len(chunks[selected_chunk]) < pos:
        skipped += len(chunks[selected_chunk])
        selected_chunk += 1
    pos -= skipped

    if len(chunks[selected_chunk]) == 1:
        del chunks[selected_chunk]
        return

    offset = chunks[selected_chunk].a
    chunks[selected_chunk].a = offset + pos
    chunks.insert(selected_chunk, Chunk(offset, offset + pos - 2))
    
    for i in range(2):
        if len(chunks[selected_chunk + (1 - i)]) <= 0:
            del chunks[selected_chunk + (1 - i)]


def perform_insert(chunks, pos, char):
    skipped = 0
    selected_chunk = 0
    while skipped + len(chunks[selected_chunk]) < pos:
        skipped += len(chunks[selected_chunk])
        selected_chunk += 1
    pos -= skipped

    if len(chunks[selected_chunk]) == 1:
        chunks.insert(selected_chunk, Chunk(char, None, True))
        return

    offset = chunks[selected_chunk].a
    new_chunk = Chunk(offset + pos - 1, chunks[selected_chunk].b)
    chunks[selected_chunk].b = offset + pos - 2
    chunks.insert(selected_chunk + 1, new_chunk)
    chunks.insert(selected_chunk + 1, Chunk(char, None, True))
        
    for i in range(3):
        if len(chunks[selected_chunk + (2 - i)]) <= 0:
            del chunks[selected_chunk + (2 - i)]

chunks1 = [Chunk(1)]
chunks2 = [Chunk(1)]

for i, working_chunk in enumerate([chunks1, chunks2]):
    while True:
        command = input().split()
        if command[0] == 'E':
#            print("==============")
            break
        if command[0] == 'D':
            pos = int(command[1])
            perform_delete(working_chunk, pos)
        if command[0] == 'I':
            pos = int(command[1])
            char = command[2]
            perform_insert(working_chunk, pos, char)
#        print(list(map(str, working_chunk)))


merged_chunks = [[chunks1[0]], [chunks2[0]]]

for i, working_chunks in enumerate([chunks1, chunks2]):
    pointer = 1
    while pointer < len(working_chunks):
        if not merged_chunks[i][-1].merge(working_chunks[pointer]):
            merged_chunks[i].append(working_chunks[pointer])
        pointer += 1


#print(list(map(str,merged_chunks[0])), list(map(str, merged_chunks[1])))

if len(merged_chunks[0]) != len(merged_chunks[1]):
    print(1)
else:
    for i in range(len(merged_chunks[0])):
        if not merged_chunks[0][i].equals(merged_chunks[1][i]):
            print(1)
            break
    else:
        print(0)

